src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -72,95 +95,152 @@ public HdfsWriter(HdfsSinkConnectorConfig connectorConfig) {
     }
   }
 
-
   public Map<TopicPartition, Long> getPreviousOffsets() {
     return offsets;
   }
 
-  public void writeRecord(TopicPartition topicPart, SinkRecord record)
-      throws CopycatException{
+  public void writeRecord(final TopicPartition topicPart, final SinkRecord record)
+      throws CopycatException, InterruptedException {
+    lock.lock();
     try {
+      recover(topicPart);
+
       RecordWriter<Long, SinkRecord> writer = getWriter(topicPart, record);
       writer.write(System.currentTimeMillis(), record);
+
       if (!offsets.containsKey(topicPart)) {
         offsets.put(topicPart, record.getOffset() - 1);
       }
       updateRecordCounter(topicPart);
+
+      if (!tasks.containsKey(topicPart)) {
+        ScheduledFuture task = timedRotateThread.schedule(new Runnable() {
+          @Override
+          public void run() {
+            lock.lock();
+            try {
+              rotate(topicPart);
+            } catch (IOException | InterruptedException e) {
+              throw new CopycatRuntimeException(e);
+            } finally {
+              lock.unlock();
+            }
+          }
+        }, rotateInterval, TimeUnit.SECONDS);
+        tasks.put(topicPart, task);
+      }
       if (shouldRotate(topicPart)) {
         rotate(topicPart);
       }
     } catch (IOException e) {
       throw new CopycatRuntimeException(e);
+    } finally {
+      lock.unlock();
     }
   }
 
-  public void close() throws IOException, CopycatException {
-    for (TopicPartition topicPart: writers.keySet()) {
+  public void close() throws IOException, CopycatException, InterruptedException {
+    timedRotateThread.shutdown();
+
+    for (TopicPartition topicPart : writers.keySet()) {
       rotate(topicPart);
     }
     writers.clear();
+
+    for (WAL wal : wals.values()) {
+      wal.close();
+    }
+    wals.clear();
+
+    tempFileNames.clear();
+    recordCounters.clear();
+    offsets.clear();
+    tasks.clear();
     if (fs != null) {
       fs.close();
     }
   }
 
-  private Map<TopicPartition, Long> readOffsets() throws IOException {
-    Map<TopicPartition, Long> offsets = new HashMap<TopicPartition, Long>();
+  private void recover(TopicPartition topicPart) throws IOException {
+    if (!recovered.contains(topicPart)) {
+      WAL wal = new WAL(conf, url, topicsDir, topicPart);
+      wal.apply();
+      wals.put(topicPart, wal);
+      recovered.add(topicPart);
+      readOffsets(topicPart);
```

Comment:
Do we need to be careful about the order of operations here? What if we add the topic partition to the `wals` map and the `recovered` set, but then reading offsets fails for some reason, we'll be left in a bad state. The next iteration will not rerun this recovery process and the offsets will be incorrect. I think we need to apply the WAL, then make sure we can read all the info we need, then atomically update our state and allow processing to continue.